staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/validation.go

```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schema
+
+import (
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+var intOrStringAnyOf = []NestedValueValidation{
+	{ForbiddenGenerics: Generic{
+		Type: "integer",
+	}},
+	{ForbiddenGenerics: Generic{
+		Type: "string",
+	}},
+}
+
+// ValidateStructural checks that s is a structural schema with the invariants:
+//
+// * structurality: both `ForbiddenGenerics` and `ForbiddenExtensions` only have zero values, with the two exceptions for IntOrString.
+// * RawExtension: for every schema with `x-kubernetes-embedded-resource: true`, `x-kubernetes-preserve-unknown-fields: true` and `type: object` are set
+// * IntOrString: for `x-kubernetes-int-or-string: true` either `type` is empty under `anyOf` and `allOf` or the schema structure is either
+//
+// 	 1) anyOf:
+//      - type: integer
+//	    - type: string
+//	 2) allOf:
+//	    - anyOf:
+//	      - type: integer
+//	      - type: string
+//	    - ... zero or more
+//
+// ValidateStructural does not check for completeness.
+func ValidateStructural(s *Structural, fldPath *field.Path) field.ErrorList {
```

Comment:
is there a reason to export ValidateStructural and ValidateStructuralCompleteness separately (validateStructural is used internally multiple times and makes sense... I'm just referring to the exported entry point(s))? when would we use one and not the other on a top-level CRD schema? I'd have a hard time explaining the difference between these to someone, and I'm concerned callers will only call one when they should call both accidentally

Reply:
good point. Have made all sub-funcs (including completeness) private and only export `ValidateStructural`.